File: box-and-box/temporal.mjs
// temporal.mjs — Temporal Arithmetic, faithful runtime (v0.4)
// The fourth rung: properties over TRAJECTORIES, not single states. A Spec is an LTL
// formula over atomic predicates on states. The core operation is `progress(φ, s)` —
// formula progression, the LTL "derivative": the residual obligation on the rest of the
// trajectory after observing state s. Monitoring is a fold of progress over the states.
//   safety   = "nothing bad ever happens"   — always(·); a violation has a finite witness.
//   liveness = "something good eventually"   — eventually(·); only failable at the horizon.
// Every linear property is (safety ∧ liveness) — Alpern & Schneider 1985. Laws T1–T8.

// ---- formula constructors (Boolean-simplifying, so residuals collapse to ⊤/⊥) ----
export const TRUE = { t: 'true' };
export const FALSE = { t: 'false' };
export const atom = (name, pred) => ({ t: 'atom', name, pred });
const isT = (f) => f.t === 'true';
const isF = (f) => f.t === 'false';

export function eq(a, b) {
  if (a === b) return true;
  if (!a || !b || a.t !== b.t) return false;
  switch (a.t) {
    case 'true': case 'false': return true;
    case 'atom': return a.name === b.name;
    case 'not': case 'next': case 'always': case 'eventually': return eq(a.a, b.a);
    case 'and': case 'or': case 'until': return eq(a.a, b.a) && eq(a.b, b.b);
    default: return false;
  }
}
export function not(f) {
  if (isT(f)) return FALSE; if (isF(f)) return TRUE;
  if (f.t === 'not') return f.a;            // ¬¬φ = φ
  return { t: 'not', a: f };
}
export function and(a, b) {
  if (isF(a) || isF(b)) return FALSE;
  if (isT(a)) return b; if (isT(b)) return a;
  if (eq(a, b)) return a;
  return { t: 'and', a, b };
}
export function or(a, b) {
  if (isT(a) || isT(b)) return TRUE;
  if (isF(a)) return b; if (isF(b)) return a;
  if (eq(a, b)) return a;
  return { t: 'or', a, b };
}
export const next = (a) => ({ t: 'next', a });
export const always = (a) => ({ t: 'always', a });          // G
export const eventually = (a) => ({ t: 'eventually', a });  // F
export const until = (a, b) => ({ t: 'until', a, b });       // U
// derived
export const gf = (a) => always(eventually(a));   // recurrence — infinitely often
export const fg = (a) => eventually(always(a));   // stabilization — eventually always
export const responds = (p, q) => always(or(not(p), eventually(q))); // p ⟹ ◇q

// ---- progress : Spec × State → Spec  (formula progression / the LTL derivative) ----
// τ ⊨ φ  iff  τ¹ ⊨ progress(φ, τ₀).   This identity is law T4.
export function progress(f, s) {
  switch (f.t) {
    case 'true': return TRUE;
    case 'false': return FALSE;
    case 'atom': return f.pred(s) ? TRUE : FALSE;
    case 'not': return not(progress(f.a, s));
    case 'and': return and(progress(f.a, s), progress(f.b, s));
    case 'or': return or(progress(f.a, s), progress(f.b, s));
    case 'next': return f.a;                                              // X φ ⇒ residual φ
    case 'always': return and(progress(f.a, s), always(f.a));            // G φ ≡ φ ∧ X G φ
    case 'eventually': return or(progress(f.a, s), eventually(f.a));     // F φ ≡ φ ∨ X F φ
    case 'until': return or(progress(f.b, s), and(progress(f.a, s), until(f.a, f.b))); // φUψ ≡ ψ ∨ (φ ∧ X(φUψ))
    default: return f;
  }
}

// ---- finite-trace closure: weak G (holds at end), strong F/U/X (fail at end) ----
function finalize(f) {
  switch (f.t) {
    case 'true': return true;
    case 'false': return false;
    case 'atom': return false;          // no state left
    case 'not': return !finalize(f.a);
    case 'and': return finalize(f.a) && finalize(f.b);
    case 'or': return finalize(f.a) || finalize(f.b);
    case 'always': return true;         // safety vacuously holds on the empty suffix
    case 'eventually': return false;    // liveness unmet by the horizon
    case 'until': return false;
    case 'next': return false;
    default: return false;
  }
}

// ---- monitor : Spec × Trajectory → verdict (+ step-by-step) ----
// online verdict: 'sat' once no extension can violate, 'vio' once none can satisfy, else 'pending'.
// final verdict over a *finished* finite trace applies the closure above.
export function monitor(f, trajectory) {
  let residual = f;
  const trace = [];
  let decidedAt = null;
  trajectory.forEach((s, i) => {
    residual = progress(residual, s);
    const v = isT(residual) ? 'sat' : isF(residual) ? 'vio' : 'pending';
    if (decidedAt === null && v !== 'pending') decidedAt = i;
    trace.push(v);
  });
  const finalSat = isT(residual) ? true : isF(residual) ? false : finalize(residual);
  return { verdict: finalSat ? 'satisfied' : 'violated', online: trace, residual, decidedAt };
}

// ---- evalDirect : independent reference semantics (finite trace), for law T4 ----
export function evalDirect(f, τ, i = 0) {
  if (i >= τ.length) { // empty suffix
    switch (f.t) {
      case 'true': return true; case 'false': return false; case 'atom': return false;
      case 'not': return !evalDirect(f.a, τ, i);
      case 'and': return evalDirect(f.a, τ, i) && evalDirect(f.b, τ, i);
      case 'or': return evalDirect(f.a, τ, i) || evalDirect(f.b, τ, i);
      case 'always': return true; case 'eventually': return false;
      case 'until': return false; case 'next': return false; default: return false;
    }
  }
  switch (f.t) {
    case 'true': return true;
    case 'false': return false;
    case 'atom': return !!f.pred(τ[i]);
    case 'not': return !evalDirect(f.a, τ, i);
    case 'and': return evalDirect(f.a, τ, i) && evalDirect(f.b, τ, i);
    case 'or': return evalDirect(f.a, τ, i) || evalDirect(f.b, τ, i);
    case 'next': return evalDirect(f.a, τ, i + 1);
    case 'always': return evalDirect(f.a, τ, i) && evalDirect(always(f.a), τ, i + 1);
    case 'eventually': return evalDirect(f.a, τ, i) || evalDirect(eventually(f.a), τ, i + 1);
    case 'until': return evalDirect(f.b, τ, i) || (evalDirect(f.a, τ, i) && evalDirect(until(f.a, f.b), τ, i + 1));
    default: return false;
  }
}

// ---- ω-words as lassos: ⟨stem⟩⟨loop⟩^ω. Direct semantics for the atomic patterns. ----
const someState = (states, p) => states.some((s) => p.pred(s));
const everyState = (states, p) => states.every((s) => p.pred(s));
export function monitorLasso(f, stem, loop) {
  // covers the headline single-atom temporal forms used in monitoring
  if (f.t === 'always' && f.a.t === 'atom') return everyState(stem, f.a) && everyState(loop, f.a);            // G p
  if (f.t === 'eventually' && f.a.t === 'atom') return someState(stem, f.a) || someState(loop, f.a);          // F p
  if (f.t === 'always' && f.a.t === 'eventually' && f.a.a.t === 'atom') return someState(loop, f.a.a);        // GF p — ∞ often
  if (f.t === 'eventually' && f.a.t === 'always' && f.a.a.t === 'atom') return everyState(loop, f.a.a);       // FG p — eventually always
  // general fallback: unroll a few loop copies and use finite semantics (sound for safety)
  return evalDirect(f, [...stem, ...loop, ...loop, ...loop]);
}

// ---- classification (coarse hint; the behavioural laws T5 are the real claim) ----
const has = (f, tags) => !!f && typeof f === 'object' && (tags.includes(f.t) || has(f.a, tags) || has(f.b, tags));
export function character(f) {
  const live = has(f, ['eventually', 'until']);
  if (f.t === 'always' && f.a && f.a.t === 'eventually') return 'liveness'; // GF
  if (!live) return 'safety';
  return has(f, ['always']) ? 'mixed' : 'liveness';
}

export function show(f) {
  switch (f.t) {
    case 'true': return '⊤'; case 'false': return '⊥'; case 'atom': return f.name;
    case 'not': return `¬${show(f.a)}`; case 'and': return `(${show(f.a)}∧${show(f.b)})`;
    case 'or': return `(${show(f.a)}∨${show(f.b)})`; case 'next': return `X${show(f.a)}`;
    case 'always': return `G${show(f.a)}`; case 'eventually': return `F${show(f.a)}`;
    case 'until': return `(${show(f.a)} U ${show(f.b)})`; default: return '?';
  }
}

export default {
  TRUE, FALSE, atom, not, and, or, next, always, eventually, until, gf, fg, responds,
  eq, progress, monitor, evalDirect, monitorLasso, character, show
};

File: box-and-box/ports/elixir/lib/box_and_box/temporal.ex
defmodule BoxAndBox.Temporal do
  @moduledoc """
  Temporal Arithmetic — faithful port of temporal.mjs (v0.4). Laws T1–T8.

  LTL over atomic predicates on states. `progress/2` is formula progression (the
  LTL derivative). Boolean-simplifying constructors collapse residuals to ⊤/⊥.
  """

  @truef %{t: "true"}
  @falsef %{t: "false"}

  def truef, do: @truef
  def falsef, do: @falsef
  def atom(name, pred), do: %{t: "atom", name: name, pred: pred}

  defp is_t(f), do: f.t == "true"
  defp is_f(f), do: f.t == "false"

  def eq(a, b) when a == b, do: true

  def eq(a, b) do
    cond do
      a == nil or b == nil or a.t != b.t ->
        false

      true ->
        case a.t do
          "true" -> true
          "false" -> true
          "atom" -> a.name == b.name
          t when t in ["not", "next", "always", "eventually"] -> eq(a.a, b.a)
          t when t in ["and", "or", "until"] -> eq(a.a, b.a) and eq(a.b, b.b)
          _ -> false
        end
    end
  end

  def not_(f) do
    cond do
      is_t(f) -> @falsef
      is_f(f) -> @truef
      f.t == "not" -> f.a
      true -> %{t: "not", a: f}
    end
  end

  def and_(a, b) do
    cond do
      is_f(a) or is_f(b) -> @falsef
      is_t(a) -> b
      is_t(b) -> a
      eq(a, b) -> a
      true -> %{t: "and", a: a, b: b}
    end
  end

  def or_(a, b) do
    cond do
      is_t(a) or is_t(b) -> @truef
      is_f(a) -> b
      is_f(b) -> a
      eq(a, b) -> a
      true -> %{t: "or", a: a, b: b}
    end
  end

  def next(a), do: %{t: "next", a: a}
  def always(a), do: %{t: "always", a: a}
  def eventually(a), do: %{t: "eventually", a: a}
  def until(a, b), do: %{t: "until", a: a, b: b}

  # derived
  def gf(a), do: always(eventually(a))
  def fg(a), do: eventually(always(a))
  def responds(p, q), do: always(or_(not_(p), eventually(q)))

  # ---- progress : Spec × State → Spec ----
  def progress(f, s) do
    case f.t do
      "true" -> @truef
      "false" -> @falsef
      "atom" -> if f.pred.(s), do: @truef, else: @falsef
      "not" -> not_(progress(f.a, s))
      "and" -> and_(progress(f.a, s), progress(f.b, s))
      "or" -> or_(progress(f.a, s), progress(f.b, s))
      "next" -> f.a
      "always" -> and_(progress(f.a, s), always(f.a))
      "eventually" -> or_(progress(f.a, s), eventually(f.a))
      "until" -> or_(progress(f.b, s), and_(progress(f.a, s), until(f.a, f.b)))
      _ -> f
    end
  end

  # ---- finite-trace closure ----
  defp finalize(f) do
    case f.t do
      "true" -> true
      "false" -> false
      "atom" -> false
      "not" -> not finalize(f.a)
      "and" -> finalize(f.a) and finalize(f.b)
      "or" -> finalize(f.a) or finalize(f.b)
      "always" -> true
      "eventually" -> false
      "until" -> false
      "next" -> false
      _ -> false
    end
  end

  # ---- monitor : Spec × Trajectory → verdict (+ step-by-step) ----
  def monitor(f, trajectory) do
    {residual, trace_rev, decided_at} =
      trajectory
      |> Enum.with_index()
      |> Enum.reduce({f, [], nil}, fn {s, i}, {res, trace, decided} ->
        res2 = progress(res, s)
        v = cond do
          is_t(res2) -> "sat"
          is_f(res2) -> "vio"
          true -> "pending"
        end
        decided2 = if decided == nil and v != "pending", do: i, else: decided
        {res2, [v | trace], decided2}
      end)

    trace = Enum.reverse(trace_rev)

    final_sat =
      cond do
        is_t(residual) -> true
        is_f(residual) -> false
        true -> finalize(residual)
      end

    %{
      verdict: if(final_sat, do: "satisfied", else: "violated"),
      online: trace,
      residual: residual,
      decided_at: decided_at
    }
  end

  # ---- eval_direct : independent reference semantics (finite trace) ----
  def eval_direct(f, tau, i \\ 0)

  def eval_direct(f, tau, i) when i >= length(tau) do
    case f.t do
      "true" -> true
      "false" -> false
      "atom" -> false
      "not" -> not eval_direct(f.a, tau, i)
      "and" -> eval_direct(f.a, tau, i) and eval_direct(f.b, tau, i)
      "or" -> eval_direct(f.a, tau, i) or eval_direct(f.b, tau, i)
      "always" -> true
      "eventually" -> false
      "until" -> false
      "next" -> false
      _ -> false
    end
  end

  def eval_direct(f, tau, i) do
    case f.t do
      "true" -> true
      "false" -> false
      "atom" -> !!f.pred.(Enum.at(tau, i))
      "not" -> not eval_direct(f.a, tau, i)
      "and" -> eval_direct(f.a, tau, i) and eval_direct(f.b, tau, i)
      "or" -> eval_direct(f.a, tau, i) or eval_direct(f.b, tau, i)
      "next" -> eval_direct(f.a, tau, i + 1)
      "always" -> eval_direct(f.a, tau, i) and eval_direct(always(f.a), tau, i + 1)
      "eventually" -> eval_direct(f.a, tau, i) or eval_direct(eventually(f.a), tau, i + 1)
      "until" ->
        eval_direct(f.b, tau, i) or
          (eval_direct(f.a, tau, i) and eval_direct(until(f.a, f.b), tau, i + 1))

      _ -> false
    end
  end

  # ---- ω-words as lassos ----
  defp some_state(states, p), do: Enum.any?(states, fn s -> p.pred.(s) end)
  defp every_state(states, p), do: Enum.all?(states, fn s -> p.pred.(s) end)

  def monitor_lasso(f, stem, loop) do
    cond do
      f.t == "always" and f.a.t == "atom" ->
        every_state(stem, f.a) and every_state(loop, f.a)

      f.t == "eventually" and f.a.t == "atom" ->
        some_state(stem, f.a) or some_state(loop, f.a)

      f.t == "always" and f.a.t == "eventually" and f.a.a.t == "atom" ->
        some_state(loop, f.a.a)

      f.t == "eventually" and f.a.t == "always" and f.a.a.t == "atom" ->
        every_state(loop, f.a.a)

      true ->
        eval_direct(f, stem ++ loop ++ loop ++ loop)
    end
  end

  # ---- classification ----
  defp has(f, tags) do
    !!f and is_map(f) and
      (Enum.member?(tags, f.t) or has(Map.get(f, :a), tags) or has(Map.get(f, :b), tags))
  end

  def character(f) do
    live = has(f, ["eventually", "until"])

    cond do
      f.t == "always" and f.a && f.a.t == "eventually" -> "liveness"
      not live -> "safety"
      has(f, ["always"]) -> "mixed"
      true -> "liveness"
    end
  end

  def show(f) do
    case f.t do
      "true" -> "⊤"
      "false" -> "⊥"
      "atom" -> f.name
      "not" -> "¬#{show(f.a)}"
      "and" -> "(#{show(f.a)}∧#{show(f.b)})"
      "or" -> "(#{show(f.a)}∨#{show(f.b)})"
      "next" -> "X#{show(f.a)}"
      "always" -> "G#{show(f.a)}"
      "eventually" -> "F#{show(f.a)}"
      "until" -> "(#{show(f.a)} U #{show(f.b)})"
      _ -> "?"
    end
  end
end
